pkg/providers/openstack/delete_instance.go

```diff
@@ -2,8 +2,40 @@ package openstack
 
 import (
 	"github.com/emc-advanced-dev/pkg/errors"
+	"github.com/emc-advanced-dev/unik/pkg/types"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/compute/v2/servers"
 )
 
 func (p *OpenstackProvider) DeleteInstance(id string, force bool) error {
-	return errors.New("not yet supportded for openstack", nil)
+	instance, err := p.GetInstance(id)
+	if err != nil {
+		return errors.New("retrieving instance "+id, err)
+	}
+	if instance.State == types.InstanceState_Running && !force {
+		return errors.New("instance "+instance.Id+" is still running. try again with --force or power off instance first", err)
+	}
+
+	clientNova, err := p.newClientNova()
+	if err != nil {
+		return err
+	}
+
+	if deleteErr := deleteServer(clientNova, instance.Id); deleteErr != nil {
+		return errors.New("failed to terminate instance "+instance.Id, deleteErr)
+	}
+
+	// Update state.
+	if err := p.state.ModifyInstances(func(instances map[string]*types.Instance) error {
+		delete(instances, instance.Id)
+		return nil
```

Comment:
I am not sure, but this seems to be a very strong pattern for updating the state (more than 30 uses). Please post this same question to upstream repo so that ilackarms will explain the logic behind.